File: 20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        if len(s) == 1:
            return False
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '{' or s[i] == '[':
                stk.append(s[i])
            elif s[i] == ')' or s[i] == '}'  or s[i] == ']':
                if stk == []:
                    return False
                t = stk.pop()
                if s[i] == ')' and t != '(' or s[i] == '}' and t != '{' or s[i] == ']' and t != '[':
                    return False
        return stk == []

File: 20.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

class Solution {
public:
  bool isValid(string s) {
    std::stack<char> st;
    for (int i = 0; i < s.size(); i++) {
      switch(s[i]) {
        case '(':
        case '[':
        case '{':
          st.push(s[i]);
          break;
        case ')':
          if (st.empty()) return false;
          if (st.top() == '(') {
            st.pop();
          } else {
            return false;
          }
          break;
        case '}':
          if (st.empty()) return false;
          if (st.top() == '{') {
            st.pop();
          } else {
            return false;
          }
          break;
        case ']':
          if (st.empty()) return false;
          if (st.top() == '[') {
            st.pop();
          } else {
            return false;
          }
          break;
      }
    }
    return st.empty();
  }
};

int main() {
  Solution s;
  bool res = s.isValid("(((({)))))");
  cout << res;
}
